src/main/java/org/mskcc/oncokb/curation/security/CustomOAuthSuccessHandler.java

```diff
@@ -57,13 +61,33 @@ public void onAuthenticationSuccess(HttpServletRequest request, HttpServletRespo
                 authorities,
                 ((OAuth2AuthenticationToken) authentication).getAuthorizedClientRegistrationId()
             );
-            SecurityContextHolder.getContext().setAuthentication(authenticationWithAuthorities);
-            super.onAuthenticationSuccess(request, response, authenticationWithAuthorities);
-        } else {
-            request.getSession().invalidate();
-            SecurityContextHolder.clearContext();
-            redirectStrategy.sendRedirect(request, response, "/logout");
+
+            // Create custom token for frontend Firebase authentication
+            String email = authenticationWithAuthorities.getPrincipal().getAttribute("email");
+            String firebaseCustomToken = null;
+            try {
+                Map<String, Object> additionalClaims = new HashMap<>();
+                // This claim will be used in Firebase rules for authorization
+                additionalClaims.put(Constants.FIREBASE_AUTHORIZED_CLAIM, true);
+                firebaseCustomToken = FirebaseAuth.getInstance().createCustomTokenAsync(email, additionalClaims).get();
+            } catch (InterruptedException | ExecutionException e) {
+                logger.error("Could not create Firebase custom token", e);
+            }
+
+            if (firebaseCustomToken != null) {
+                Map<String, Object> additionalDetails = new HashMap<>();
+                additionalDetails.put(Constants.FIREBASE_CUSTOM_TOKEN, firebaseCustomToken);
+                authenticationWithAuthorities.setDetails(additionalDetails);
+
+                SecurityContextHolder.getContext().setAuthentication(authenticationWithAuthorities);
+                super.onAuthenticationSuccess(request, response, authenticationWithAuthorities);
+                return;
```

Comment:
I assume the user will be logout if they don't have the access to the firebase? That's not the intended use case.

Reply:
The user will be logged out if a firebase token cannot be created due to error. Should we still let the user login, but on the frontend we would need to make some components that rely on firebase unavailable?

Reply:
Yes, not all of users of curation platform will have the access to firebase.

Reply:
We could create a ROLE_FIREBASE to differentiate.